source/mainnet/smart-contracts/onboarding-guide-solana-developers/overview.rst

```diff
@@ -0,0 +1,112 @@
+
+Solana programs take a list of accounts as input.
+Some of these accounts can be used to store persistent data, that is, data that is stored after the contract invocation is completed.
+Concordium uses smart contract states for that purpose.
+Each contract can write to its own state directly.
+Modifying the state of other smart contracts is only possible by calling their entrypoints.
+Smart contracts cannot read/write arbitrary data from/to user accounts, but they can transfer CCD they own to any user account and other smart contracts.
+Note that CCD cannot be transferred to the smart contract balance without calling a specific entrypoint or a :ref:`fallback entrypoint<fallback-entrypoints>`.
+The smart contract logic determines whether to accept CCD or not.
+
+Example
+=======
+
+This example demonstrates the difference in how the persistent state is handled.
+
+Solana
+------
+
+The example code is taken from `example-helloworld <https://github.com/solana-labs/example-helloworld>`_ and adjusted.
+
+.. code-block:: rust
+
+        // Define the type of state stored in accounts
+        #[derive(BorshSerialize, BorshDeserialize, Debug)]
+        pub struct GreetingAccount {
+            pub counter: u32,
+        }
+
+        ...
+
+        pub fn process_instruction(
+            program_id: &Pubkey, // Public key of the program account
+            accounts: &[AccountInfo], // A slice of accounts to operate on
+            _instruction_data: &[u8], // Instructions, ignored in this example
+        ) -> ProgramResult {
+
+            let accounts_iter = &mut accounts.iter();
+
+            // Get the account from the input to store the state
+            let account = next_account_info(accounts_iter)?;
+
+            // The account must be owned by the program in order to modify its data
+            if account.owner != program_id {
+                msg!("Greeted account does not have the correct program id");
+                return Err(ProgramError::IncorrectProgramId);
+            }
+
+            // Increment and store the number of times the account has been greeted
+            let mut greeting_account = GreetingAccount::try_from_slice(&account.data.borrow())?;
+            greeting_account.counter += 1;
+            greeting_account.serialize(&mut &mut account.data.borrow_mut()[..])?;
+
+            Ok(())
+        }
+
+Concordium
+----------
+
+.. code-block:: rust
+
+    // Define the type of state for the contract
+    #[derive(Serialize, SchemaType, Clone)]
+    pub struct MyState {
+        counter: u32
+    }
+
+   #[receive(
+        contract = "example",
+        name = "some_receive",
+        mutable,
+    )]
+    fn some_receive<S: HasStateApi>(
+        _ctx: &impl HasReceiveContext,
+        host: &mut impl HasHost<MyState, StateApiType = S>,
+    ) -> Result<(), Error> {
+
+        // Load the contract state; increment and store the counter
+        host.state_mut().counter += 1;
+
+        Ok(())
+    }
```

Comment:
This does not look that different. 

On concordium the state is a key-value store, which can also be thought of as a file system. There is a notion of directory that arises from iterators.

The main distinction seems to be that on Solana you supply the keys/files to the call directly, and each file has its own owner potentially. Whereas on Concordium a contract has its own file system that it owns directly.

But this is not really a big distinction and is more a point of view. Since you can just have multiple contract instances.

Reply:
Yes, I understand it in the same way. Maybe it's not a big distinction, but still affects how the code is organised.
For me, it looks a bit like object-oriented vs procedural.